### ISSUE TITLE
Using <= comparison operator for strings in F# inside a query (linq-like) syntax 

### ISSUE THREAD

**Issue by giuliohome** [dotnet/fsharp#9691]
**Repro steps**

I'm reporting the issue here [after a first analysis on SO, as suggested there](https://stackoverflow.com/questions/62874859/using-the-less-than-comparison-operator-for-strings-in-f?noredirect=1#comment111215563_62874859).
I'm trying to write a F# query as follows

```
let extract = 
    query {
        for trade in Schema.Trade do
        where (trade.Year + trade.Month <= "202007")
        take 100
        select (trade)
    }
```

The problem is that both `Year` and `Month` are `string`.

I see an exception coming from Microsoft.FSharp.Linq.QueryModule saying that

`LessThanOrEqual` is not defined for the types `System.String` and `System.String`.

Why not? Is there a different operator?

To give you more context, the query is running inside SQLProvider and, if I write only `=` instead of `<=`, the auto-generated sql already translates the + into a correct
```
CONCAT(`trade`.`year`, `trade`.`month`)
```
So that part is not an issue on the SQLProvider side.

Let me rephrase what I think is happening under the cover. Sorry, this is just my fealing about the analysis of the problem, I don't want to confuse things. Anyway, I guess that something like the following is happening (this is C# interactive)

```
> Expression<Func<String>> ex1 = () => "202007";;
> Expression<Func<String>> ex2 = () => "202006";;
> Expression.GreaterThan(ex1.Body, ex2.Body);;
System.InvalidOperationException: L'operatore binario GreaterThan non è definito per i tipi 'System.String' e 'System.String'.
  + System.Linq.Expressions.Expression.GetUserDefinedBinaryOperatorOrThrow(System.Linq.Expressions.ExpressionType, string, System.Linq.Expressions.Expression, System.Linq.Expressions.Expression, bool)
  + System.Linq.Expressions.Expression.GreaterThan(System.Linq.Expressions.Expression, System.Linq.Expressions.Expression, bool, System.Reflection.MethodInfo)
```

**Expected behavior**

Both F# language and the underlying SQL can natively manage a string comparison, so I'm trying to get it rendered to sql.
Aside from the sql transalation (that is in charge of the type provider and that cannot even start in this case, due to the preliminary exception on the F# level), I would expect that, first of all, the expression generated from the F# query should be expressed as a correct AST in linq-like format without any exception.

**Actual behavior**

This is instead the real, full stack trace from the above said F# program.

```
</ExceptionType><Message>L'operatore binario LessThanOrEqual non è definito per i tipi 'System.String' e 'System.String'.</Message><StackTrace>   in System.Linq.Expressions.Expression.GetUserDefinedBinaryOperatorOrThrow(ExpressionType binaryType, String name, Expression left, Expression right, Boolean liftToNull)
   in System.Linq.Expressions.Expression.LessThanOrEqual(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext@512-7.Invoke(Tuple`2 tupledArg)
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 512
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 734
   in Microsoft.FSharp.Primitives.Basics.List.mapToFreshConsTail[a,b](FSharpList`1 cons, FSharpFunc`2 f, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 241
   in Microsoft.FSharp.Primitives.Basics.List.map[T,TResult](FSharpFunc`2 mapping, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 252
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprsToLinq(ConvEnv env, FSharpList`1 es) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 820
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 642
   in Microsoft.FSharp.Primitives.Basics.List.map[T,TResult](FSharpFunc`2 mapping, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 250
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprsToLinq(ConvEnv env, FSharpList`1 es) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 820
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 642
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 734
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.EvaluateQuotation(FSharpExpr e) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 844
   in Microsoft.FSharp.Linq.QueryModule.EvalNonNestedInner(CanEliminate canElim, FSharpExpr queryProducingSequence) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Query.fs:riga 1823
   in Microsoft.FSharp.Linq.QueryModule.clo@1926-1.Microsoft-FSharp-Linq-ForwardDeclarations-IQueryMethods-Execute[a,b](FSharpExpr`1 q) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Query.fs:riga 1928
   in Program.main(String[] argv) in C:\FunctionalCmd\Program.fs:riga 23</StackTrace><ExceptionString>System.InvalidOperationException: L'operatore binario LessThanOrEqual non è definito per i tipi 'System.String' e 'System.String'.
   in System.Linq.Expressions.Expression.GetUserDefinedBinaryOperatorOrThrow(ExpressionType binaryType, String name, Expression left, Expression right, Boolean liftToNull)
   in System.Linq.Expressions.Expression.LessThanOrEqual(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext@512-7.Invoke(Tuple`2 tupledArg)
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 512
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 734
   in Microsoft.FSharp.Primitives.Basics.List.mapToFreshConsTail[a,b](FSharpList`1 cons, FSharpFunc`2 f, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 241
   in Microsoft.FSharp.Primitives.Basics.List.map[T,TResult](FSharpFunc`2 mapping, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 252
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprsToLinq(ConvEnv env, FSharpList`1 es) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 820
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 642
   in Microsoft.FSharp.Primitives.Basics.List.map[T,TResult](FSharpFunc`2 mapping, FSharpList`1 x) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\local.fs:riga 250
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprsToLinq(ConvEnv env, FSharpList`1 es) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 820
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 642
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 734
   in Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.EvaluateQuotation(FSharpExpr e) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Linq.fs:riga 844
   in Microsoft.FSharp.Linq.QueryModule.EvalNonNestedInner(CanEliminate canElim, FSharpExpr queryProducingSequence) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Query.fs:riga 1823
   in Microsoft.FSharp.Linq.QueryModule.clo@1926-1.Microsoft-FSharp-Linq-ForwardDeclarations-IQueryMethods-Execute[a,b](FSharpExpr`1 q) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\Query.fs:riga 1928
   in Program.main(String[] argv) in C:\FunctionalCmd\Program.fs:riga 23</ExceptionString></Exception></TraceRecord>
Eccezione non gestita di tipo 'System.InvalidOperationException' in FSharp.Core.dll
L'operatore binario LessThanOrEqual non è definito per i tipi 'System.String' e 'System.String'.
```

**Known workarounds**

In my specific case I think I can detour the issue by a

```
let extract = 
    let limitDate = System.DateTime(2020,08,01)
    query {
        for trade in Schema.Trade do
        (trade.Mydate < lmitDate ) 
        take 100
        select (trade)
    }
```

 because I think I have such a column trade.Mydate in `TIMESTAMP` format from the data lake , but, generally speaking, I could receive a requirement for a `string` field..

this also seems to work as well

```
let extract = 
    query {
        for trade in Schema.Trade do
        where (Int32.Parse trade.Year < 2020 || 
          (Int32.Parse trade.Year = 2020 && Int32.Parse trade.Month <= 7))
        take 100
        select (trade)
    }
```
but again, from a functional standpoint, the requirement is different (and it's also funny/odd that one has to use `Int32.Parse` instead of `int`, because `int` throws a  "no coercion operator between `string` and `int` `at System.Linq.Expressions.Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)` etc...") and such a conversion from `string` to `int` appears really unneeded  from the sql perspective (and from a general point of view)

**Related information**

Provide any related information (optional):

* Operating system Windows 64 bit
* .NET Runtime kind: Both .NET Framework 4.8 and .NET Core 3.1
* Editing Tools (e.g. Visual Studio Version, Visual Studio): Visual Studio 2019

**Comment by giuliohome** [comment]
Or maybe has it to do with SqlRuntime.Patterns.fs of SQLProvider repo (follow my mention above)?

**Comment by abelbraaksma** [comment]
> LessThanOrEqual is not defined for the types System.String and System.String.

I'd assume this to work, as string implements `IComparable`. And any database supports this too.

**Comment by giuliohome** [comment]
> > LessThanOrEqual is not defined for the types System.String and System.String.
> 
> I'd assume this to work, as string implements `IComparable`. And any database supports this too.


@abelbraaksma  I agree with you **100%**, this **must** work.

But... the problem is _what_ is going wrong...

Now I suspect that it is the way the linq AST is formed that is wrong...
And I'm afraid that the linq AST is not generated here in this repo but in the type provider repo..

The basis for my idea is the literal interpretation of the error: if you follow the answer on SO you can read an interesting phrase

>The problem, of course, is that there is no String.GreaterThan method. The usual answer is to use String.CompareTo(),

in the C# equivalent question linked inside the first line of the answer

**Comment by cartermp** [comment]
FWIW the query translation is super straightforward, it ultimately just constructs a binary operation on two arguments with the `LessThanOrEqual` Expression API, and it's inside there that the exception is raised. Not sure if there's much to resolve on the FSharp.Core end of things.

**Comment by cartermp** [comment]
Tagging as a bug for now since it's not obvious what's going on, but i'm not clear on if this is actually an issue with expression translation

**Comment by giuliohome** [comment]
Well, it is super straightforward indeed but the transition into the
LessThanOrEqual appears wrong somehow: it looks like it is not the right
translation in the context of strings. What the human language translators
would call a false friend.
But before diving into this matter, are you sure that Fsharp Core is doing
the translation here? I was in doubt between the Core and the type
provider...
Sorry for brevity, I'm writing through email on my phone...

**Comment by abelbraaksma** [comment]
> Sorry for brevity, I'm writing through email on my phone...

Just a tip, I find it much easier to type from phone on the website, which gives a good notifications experience and interface for mobile.

**Comment by giuliohome** [comment]
I'm on my laptop now. Anyway, I'm looking at C# and I see that error in pure C# linq.
But the point is that in C# I can't do "F" > "B" while my first question here about F# is: why are you reverting back to C# linq?

**Comment by giuliohome** [comment]
I mean, the point is that from F# interactive I get
```
> "F" > "B";;
val it : bool = true
```
but then you fall back into C# linq as far as I can see

**Comment by giuliohome** [comment]
>  But before diving into this matter, are you sure that Fsharp Core is doing the translation here? I was in doubt between the Core and the type provider... Sorry for brevity, I'm writing through email on my phone...
I was wrong this is F# Core. Here it is the repro without the type provider

```fsharp
> let qtest = test.AsQueryable();;
val qtest : IQueryable<string>

and you get the exception now
> let qry = query {
    for s in qtest do
    where (s >= "b")
    select s
    };;
```

**Comment by giuliohome** [comment]
And I think it is a bug in F# language because 

```fsharp
let qry = query {
    for s in test do // notice that test is the enumerable  not iqueryable qtest
    where (s >= "b")
    select s
    };;
val qry : seq<string>

> qry |> Seq.toArray;;
val it : string [] = [|"b"; "c"|]
```

**Comment by giuliohome** [comment]
Let me test if the following is working from the sql type provider...

```
let qry = query {
    for s in qtest do
    where (s.CompareTo("b") >= 0)
    select s
    };;
```

**Comment by giuliohome** [comment]
Tested now: `CompareTo` is an unsupported expression from SQL Type Provider, so eihter you will accept it as a bug and fix it in F# Core with `>=` (which I doubt because I see that the actual linq expression implementation is from C#) or SQL Type Provider will support `CompareTo`  (which I also doubt given the status of that repository in terms of supported features) or (and that seems the only realistic option) I won't use F# + SQLType Provider in that project of mine.

**Comment by abelbraaksma** [comment]
> but then you fall back into C# linq as far as I can see

Is it "C# Linq" or just basically the existing LINQ from the BCL? 

> which I doubt because I see that the actual linq expression implementation is from C#

I don't think so, as @cartermp mentioned, this uses `LessThenOrEqual`, which is part of the BCL (and as such part of .NET Core, Framework etc), see https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression.lessthanorequal?view=netcore-3.1. I doubt there's anything C#-specific about this implementation.

> `// notice that test is the enumerable  not iqueryable qtest`

This makes sense, because the operation is a `for .. in .. do` expression, which operates on `IEnumerable`. But I've not enough knowledge of the internals of the `query` CE, which is, afaik, a thin wrapper over the LINQ statements.

I also noticed that the following works using your code:

```
let qry = query {
    for s in qtest do
    where (s.CompareTo("b") > 0)
    select s
    };;
```

But as you mentioned, not with the SQL type provider.

> I won't use F# + SQLType Provider in that project of mine.

I don't know your use case, but on Slack there have been quite active discussions on several other libraries that integrate well with F# w.r.t. databases. 

This article seems to share your concerns and shows a way to use Dapper: https://nozzlegear.com/blog/using-dapper-with-fsharp

**Comment by giuliohome** [comment]
The formal way to say it is probably using the BCL Linq term instead of my informal C# Linq, but I'm still afraid you can't find the needed method info for the binary string comparison returning a bool there in the BCL. Anyway I'm happy to be wrong if you find a fix.

As a side note, Dapper does absolutely nothing of what I really want, unfortunately. I don't like writing sql for trivial crud with dozens of columns and when I have to use linq2db or vanilla ado or odbc I prefer C#

**Comment by abelbraaksma** [comment]
> Anyway I'm happy to be wrong if you find a fix.

I'm probably not the right person. The reason I pointed out that it's likely a BCL problem, esp if it's not possible with C# either, is that we should move the bug to the runtime github repo. 

But I may misunderstand this, as in part I think you say it doesn't work on either C# or F#, but your last comment makes me think it's possible in C#.

If possible, can you share what did, or did not work in C#? Maybe we need to extend the F# query syntax, or maybe there's an easy fix, I'm really not sure.

**Comment by giuliohome** [comment]
Before I rephrase my last comment let me say that it seems impossible in both C# and F# but for the latter it sounds more "inconsistent" because it implements a `>` for strings.

The solution could be - at the level of the linq package -  for both F# and C# to use linq2db: of course it is perfectly usable also from F# and, if I remember well, it implements the `CompareTo` in the auto-generation of sql (that can be quite useful for DRY in typical business scenarios). 

I mixed a contrived opinion of mine when I added that (let me rephrase it) the value added of functional programming a db layer in F# is (partially) lost when you write code perfectly equivalent 1 to 1 to C#.

**Comment by giuliohome** [comment]
So my point is that linq2db is perfectly fine (see my previous comment, edited) to solve this issue (in C# style of programming and with C# Visual Studio support). I initially preferred F# + SqlProvider for this project because it seemed to have an already available [generic obdc provider](https://github.com/fsprojects/SQLProvider/issues/686) but it turned out it was not true and IMO at this point for such a project it would be better to work with linq2db and [implement the odbc provide there as they suggest](https://github.com/linq2db/linq2db/issues/2335#issuecomment-656104866) in C# also because it is more robust in terms of compile time check (even if F# is also usable, somehow...)

**Comment by cartermp** [comment]
Will close this out - not a bug, since we're emitting the correct call based on the underlying type. It's unfortunate that IQueryable has this behavior though - it does sort of violate the law of least surprise.
